keras_retinanet/preprocessing/generator.py

```diff
@@ -43,6 +43,7 @@ def __init__(
         image_min_side=800,
         image_max_side=1333,
         transform_parameters=None,
+        shapes_callback=None,
```

Comment:
One last remark (probably): I'm not a fan of the name `shapes_callback`. Suffixing with `_callback` is pretty much a form of Hungarian notation.

I'd prefer a name like `calculate_shapes`, although that still doesn't tell us which shapes. So perhaps something along the lines of `calculate_pyramid_shapes` or just `pyramid_shapes`? Other opinions welcome too.

Reply:
I prefer `compute` over `calculate` but just `pyramid shapes` seems ok as well.

Reply:
I agree that `shapes_callback` sounds a bit ambiguous. If I'd have to choose between `pyramid_shapes` and `compute_pyramid_shapes` I'd choose the latter, since `pyramid_shapes` sounds like it's a list of shapes, not a functor.

Reply:
Hmm, I was thinking a bit more. In the end we only use this functor to ~calculate~ compute the anchors, right? So what if we just pass a functor to compute the anchors directly? It would remove quite a few parameters and give the option to modify the anchors.

Reply:
We could pass `anchor_targets_bbox` as a functor. On the other hand I'm not sure where to place the `compute_pyramid_shapes` option in that case. What do you think?

Reply:
That would then be internal to the `compute_anchors` functor.

Reply:
I apologize, my question was very vague. What I had in mind is that the current call sequence is `Generator.anchor_targets => anchor_targets_bbox => anchors_for_shape => shapes_callback`. If `anchor_targets_bbox` replaces `shapes_callback` in the generator constructor, we still need to propagate `shapes_callback` somehow, or create two different versions of `anchor_targets_bbox` (and implicitly also two versions of `anchors_for_shape`), but the latter would lead to a lot of duplicated code.

Reply:
I think what @de-vri-es proposes is that there is a function that wraps `anchor_targets_bbox`. The wrapper would take a functor to compute pyramid shapes, or it will use the default if none are provided. That way the backbones that require the shapes to be computed based on the model can use this wrapper to create a version of `anchor_targets_bbox` that computes shapes using a functor, while the other backbones "guess" the shapes (default behaviour).

The advantage of this approach is that the generator doesn't need to pass on the `compute_pyramids_shapes` to `anchor_targets_bbox` and it doesn't need to have an attribute for the functor. The generator has no need for the pyramid shapes, only indirectly because the anchors function needs it. Does that clear things up?

Reply:
I was actually thinking about just the anchors, not the anchor targets. But the anchor targets are probably even better.